src/cryptonote_core/service_node_quorum_cop.cpp

```diff
@@ -96,18 +96,17 @@ namespace service_nodes
       }
     }
 
-    if (check_checkpoint_obligation && info.proof.num_checkpoint_votes_expected >= CHECKPOINT_MIN_QUORUMS_NODE_MUST_VOTE_IN_BEFORE_DEREGISTER_CHECK)
+    if (check_checkpoint_obligation)
     {
       proof_info const &proof = info.proof;
-      // NOTE: We can receive more votes than expected, say, the nodes are in the quorums newer than the obligation check height
-      if (proof.num_checkpoint_votes_received < proof.num_checkpoint_votes_expected)
+      int num_votes           = 0;
+      for (bool voted : proof.votes)
+        num_votes += voted;
+
+      if (num_votes <= CHECKPOINT_MAX_MISSABLE_VOTES)
```

Comment:
Right. Yeah actually I had it so that you are still included in checkpoint quorums when decommissioned, but you could be unlucky and just not included in any quorums and won't ever catch up in time before your credits run out which doesn't seem like it's fair.

So, now on decommission I reset the seen votes to all true and we just rely on the uptime proof to recommission.

Reply:
How likely is it that you go down and miss the required number of votes to trigger a decommission within 2 hours?  If this is something that isn't all that rare, you could end up going down, getting decommissioned because of the missing votes, then immediately getting recommissioned because your uptime is still less than 2 hours, then a few minutes later after the 2 hour mark, getting deregistered because you burned your decomm credit in the first short decomm/recomm round.

If that is likely to happen sometimes, we might be able to address that case if we imposed some sort of mandatory minimum 2-hour decommission when your decomm happens because of a checkpoint vote failure (i.e. long enough to get you into uptime quorum failures if the problem is indeed that your node has just gone down).

Reply:
20 nodes per block are checked for missing a vote compared to the obligations quorum max(1% of the network, 50 nodes). I was under the impression that the credit would carry over, but that doesn't look the case with active_since_height getting reset on recommission.

Yeah ok. I've thought about it a bit and there doesn't seem a nice way to handle this except compounding on extra rules onto the decommissioning phase as you've suggested. 2 hours seems a bit long, and that's because we would be trying to rely on the uptime test, to test the node.

An inbetween compromise might be that if you are decommissioned because of missing those checkpoint votes, the earliest you can be recommissioned is the earliest of (30 mins, next N checkpoint votes that pushes you over the threshold non-decommissioning threshold). 

Also, highlights the issue with well, service nodes that are decommissioned cannot be recommissioned unless they are in the obligation quorum which is max(1% of the network/50 nodes). This suggests that we want to use the obligation quorum to decommission nodes only, but every block we should be checking every node for recommissioning?

Reply:
Recommissioning isn't a huge issue because if you don't get checked it's just a delay before reactivation - you don't get deregistered unless you fail another uptime check *and* the time has run out.  But if you come back online by then, even if your test gets delayed, you'll get a recommission instead of a deregistration.

(The probabilities here: you've got about a 70% of being tested within 120 blocks once we hit the 1% testing threshold, which we won't hit until we have 5000 nodes--but by that point it will *also* be a very small fraction of the reward cycle.  At our current count of 654, you've got about a 90% chance of being retested within 30 blocks).